### PR TITLE
Additional code coverage in System.Text.Json

### PR DESCRIPTION
**Pull request by alanisaac**
Part of #32341

**Note 1**

In order to get coverage on the following lines, I added a missing test case using `JsonDocument.Parse` with a large, invalid base64 string _without_ an escaped character.  The test data previously used `[InlineData(...)]` attributes where the large strings were generated from the `[InlineData(null)]` case, using `null` as a sentinel value.  I weighed a few different options for how to add this new case, and eventually settled on moving the test data to its own class and using `[MemberData(...)]` instead to simplify the tests themselves.  Certainly open to suggestions! 

https://github.com/dotnet/runtime/blob/11b37df0e50e45465b2071b191c365d84e8932ac/src/libraries/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Unescaping.cs#L162-L166

**Note 2**
There was a missing case in cyclic tests that validates that `EffectiveMaxDepth` is set to 64 when `MaxDepth` is set to 0 (lines).  I ended up splitting `WriteCyclicFail`, which had two logical subgroups into `WriteCyclic` and `WriteCyclicFail`.

https://github.com/dotnet/runtime/blob/527adf211a45046876d680480e45131b5334fcf6/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs#L203

### CODE REVIEW

File: src/libraries/System.Text.Json/tests/Serialization/ReferenceHandlingTests.Deserialize.cs [hunk 1]
```diff
@@ -910,6 +910,16 @@ public static void ThrowOnStructWithReference()
             Assert.Equal("$[1].$ref", ex.Path);
             Assert.Contains($"'{typeof(EmployeeStruct)}'", ex.Message);
         }
+
+        [Theory]
+        [InlineData(@"{""$iz"": ""1""}")]
+        [InlineData(@"{""$rez"": ""1""}")]
+        [InlineData(@"{""$valuez"": []}")]
+        public static void InvalidMetadataPropertyNameWithSameLengthIsNotRecognized(string json)
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/tests/Serialization/ReferenceHandlingTests.Deserialize.cs]: Nice!

File: src/libraries/System.Text.Json/tests/Serialization/ReferenceHandlingTests.Deserialize.cs [hunk 2]
```diff
@@ -910,6 +910,16 @@ public static void ThrowOnStructWithReference()
...
+        {
+            JsonException ex = Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<Employee>(json, s_deserializerOptionsPreserve));
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/tests/Serialization/ReferenceHandlingTests.Deserialize.cs]: Do we want to assert anything about the returned `JsonException` itself and its properties?
If not:
```suggestion
            Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<Employee>(json, s_deserializerOptionsPreserve));
```

File: src/libraries/System.Text.Json/tests/Serialization/CyclicTests.cs [hunk 3]
```diff
@@ -21,35 +21,43 @@ public static void WriteCyclicFailDefault()
...
-        [InlineData(1)]
-        [InlineData(2)]
-        [InlineData(10)]
-        [InlineData(70)]
-        public static void WriteCyclicFail(int depth)
+        [InlineData(1, 2)]
+        [InlineData(2, 3)]
+        [InlineData(10, 11)]
+        [InlineData(70, 71)]
+        [InlineData(10, 0)]
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/tests/Serialization/CyclicTests.cs]: nit: Add a comment here that 0 (or default) max depth is treated as 64. Something like this (feel free to reword):
```suggestion
        [InlineData(10, 0)] // 0 (or default) max depth is treated as 64.
```

File: src/libraries/System.Text.Json/tests/Serialization/CyclicTests.cs [hunk 4]
```diff
@@ -21,35 +21,43 @@ public static void WriteCyclicFailDefault()
...
-                string expectedPath = "$" + string.Concat(Enumerable.Repeat(".Parent", depth - 1));
-                Assert.Contains(expectedPath, ex.Path);
-                Assert.Contains(depth.ToString(), ex.ToString());
-            }
+        [Theory]
+        [InlineData(1, 1, 0)]
+        [InlineData(2, 2, 1)]
+        [InlineData(10, 10, 9)]
+        [InlineData(70, 70, 69)]
+        [InlineData(70, 0, 63)]
```
Comment (ahsonkhan) [inline src/libraries/System.Text.Json/tests/Serialization/CyclicTests.cs]: Same nit: comment that 0 is special and means 64.

**Review by ahsonkhan — APPROVED** [review]
Some nits. But looks great.

File: src/libraries/System.Text.Json/tests/Serialization/ReferenceHandlingTests.Deserialize.cs (hunk 2 above)
Comment (alanisaac) [reply src/libraries/System.Text.Json/tests/Serialization/ReferenceHandlingTests.Deserialize.cs]: Good point -- added validation of the expected path in 9c0af54

File: src/libraries/System.Text.Json/tests/Serialization/CyclicTests.cs (hunk 3 above)
Comment (alanisaac) [reply src/libraries/System.Text.Json/tests/Serialization/CyclicTests.cs]: fixed in 70835c4

File: src/libraries/System.Text.Json/tests/Serialization/CyclicTests.cs (hunk 4 above)
Comment (alanisaac) [reply src/libraries/System.Text.Json/tests/Serialization/CyclicTests.cs]: fixed in 70835c4

File: src/libraries/System.Text.Json/tests/JsonBase64TestData.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            yield return new object[] { "\"++++\"" };
+            yield return new object[] { GenerateRandomValidLargeString() };
+        }
+
+        public static IEnumerable<object[]> InvalidBase64Tests()
+        {
+            yield return new object[] { "\"ABC===\"" };
+            yield return new object[] { "\"ABC\"" };
+            yield return new object[] { "\"ABC!\"" };
+            yield return new object[] { GenerateRandomInvalidLargeString(true) };
```
[the remark this answers is not in the document]
Comment (alanisaac) [reply src/libraries/System.Text.Json/tests/JsonBase64TestData.cs]: fixed in e4ba4ce

File: src/libraries/System.Text.Json/tests/JsonBase64TestData.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
+﻿using System.Collections.Generic;
```
[the remark this answers is not in the document]
Comment (alanisaac) [reply src/libraries/System.Text.Json/tests/JsonBase64TestData.cs]: fixed in 6c8ca54

**Comment by alanisaac** [comment]
@ahsonkhan I found that there was another set of tests for Base64 strings using the exact same test data.  I thought it would be a good idea to reuse it to reduce LoC, similar to other data sets like `ValidISO8601Tests`.  See 8c893c0

**Review by ahsonkhan — APPROVED** [review]

File: src/libraries/System.Text.Json/tests/JsonBase64TestData.cs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+            yield return new object[] { GenerateRandomInvalidLargeString(true) };
+            yield return new object[] { GenerateRandomInvalidLargeString(false) };
+        }
+
+        private static string GenerateRandomValidLargeString()
+        {
+            var random = new Random(42);
+            var charArray = new char[502];
+            charArray[0] = '"';
+            for (int i = 1; i < charArray.Length; i++)
```
[the remark this answers is not in the document]
Comment (alanisaac) [reply src/libraries/System.Text.Json/tests/JsonBase64TestData.cs]: fixed in d09874d

File: src/libraries/System.Text.Json/tests/JsonBase64TestData.cs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+            yield return new object[] { "\"ABC\"" };
+            yield return new object[] { "\"ABC!\"" };
+            yield return new object[] { GenerateRandomInvalidLargeString(true) };
+            yield return new object[] { GenerateRandomInvalidLargeString(false) };
+        }
+
+        private static string GenerateRandomValidLargeString()
+        {
+            var random = new Random(42);
+            var charArray = new char[502];
```
[the remark this answers is not in the document]
Comment (alanisaac) [reply src/libraries/System.Text.Json/tests/JsonBase64TestData.cs]: added explanation in b9b42e1

**Review by layomia — APPROVED** [review]
Thanks!

**Pull request merged**